### PR TITLE
feat(eval/rollouts): write benchmark tests for evaluations

### PR DESCRIPTION
**Pull request by yquansah**
Write benchmark tests for evaluation v1 and variant and boolean rollout evaluation.

Completes FLI-470

### CODE REVIEW

**Comment by yquansah** [comment]
This is from one run of the benchmark test:

```bash
Running tool: /opt/homebrew/bin/go test -benchmem -run=^$ -bench ^Benchmark_EvaluationV1AndV2$ go.flipt.io/flipt/internal/storage/sql

goos: darwin
goarch: arm64
pkg: go.flipt.io/flipt/internal/storage/sql
Benchmark_EvaluationV1AndV2/evaluation-v1-flag_037-10         	     558	   2170829 ns/op	  162456 B/op	    6388 allocs/op
Benchmark_EvaluationV1AndV2/variant-evaluation-flag_037-10    	     556	   2160445 ns/op	  162527 B/op	    6389 allocs/op
Benchmark_EvaluationV1AndV2/evaluation-v1-flag_001-10         	     555	   2148621 ns/op	  162424 B/op	    6388 allocs/op
Benchmark_EvaluationV1AndV2/variant-evaluation-flag_001-10    	     560	   2138909 ns/op	  162531 B/op	    6389 allocs/op
Benchmark_EvaluationV1AndV2/evaluation-v1-flag_028-10         	     552	   2156414 ns/op	  162438 B/op	    6388 allocs/op
Benchmark_EvaluationV1AndV2/variant-evaluation-flag_028-10    	     560	   2158400 ns/op	  162529 B/op	    6389 allocs/op
Benchmark_EvaluationV1AndV2/evaluation-v1-flag_041-10         	     559	   2139186 ns/op	  162442 B/op	    6388 allocs/op
Benchmark_EvaluationV1AndV2/variant-evaluation-flag_041-10    	     558	   2140845 ns/op	  162519 B/op	    6389 allocs/op
Benchmark_EvaluationV1AndV2/evaluation-v1-flag_041#01-10      	     553	   2146814 ns/op	  162445 B/op	    6388 allocs/op
Benchmark_EvaluationV1AndV2/variant-evaluation-flag_041#01-10 	     556	   2155168 ns/op	  162548 B/op	    6389 allocs/op
Benchmark_EvaluationV1AndV2/evaluation-v1-flag_041#02-10      	     550	   2173206 ns/op	  162424 B/op	    6388 allocs/op
Benchmark_EvaluationV1AndV2/variant-evaluation-flag_041#02-10 	     555	   2161792 ns/op	  162553 B/op	    6389 allocs/op
Benchmark_EvaluationV1AndV2/evaluation-v1-flag_014-10         	     556	   2186956 ns/op	  162450 B/op	    6388 allocs/op
Benchmark_EvaluationV1AndV2/variant-evaluation-flag_014-10    	     556	   2150104 ns/op	  162529 B/op	    6389 allocs/op
Benchmark_EvaluationV1AndV2/evaluation-v1-flag_046-10         	     561	   2147822 ns/op	  162446 B/op	    6388 allocs/op
Benchmark_EvaluationV1AndV2/variant-evaluation-flag_046-10    	     558	   2163105 ns/op	  162513 B/op	    6389 allocs/op
Benchmark_EvaluationV1AndV2/evaluation-v1-flag_021-10         	     542	   2186103 ns/op	  162436 B/op	    6388 allocs/op
Benchmark_EvaluationV1AndV2/variant-evaluation-flag_021-10    	     556	   2157712 ns/op	  162529 B/op	    6389 allocs/op
Benchmark_EvaluationV1AndV2/evaluation-v1-flag_001#01-10      	     555	   2146962 ns/op	  162434 B/op	    6388 allocs/op
Benchmark_EvaluationV1AndV2/variant-evaluation-flag_001#01-10 	     555	   2166358 ns/op	  162534 B/op	    6389 allocs/op
Benchmark_EvaluationV1AndV2/boolean-evaluation-flag_boolean-10         	    2116	    553998 ns/op	   30860 B/op	     678 allocs/op
Benchmark_EvaluationV1AndV2/boolean-evaluation-another_boolean_flag-10 	    2155	    557482 ns/op	   30892 B/op	     678 allocs/op
PASS
ok  	go.flipt.io/flipt/internal/storage/sql	35.499s
```

File: internal/storage/sql/evaluation_test.go [hunk 1]
```diff
@@ -682,3 +692,95 @@ func (s *DBTestSuite) TestGetEvaluationRollouts_NonDefaultNamespace() {
...
+func Benchmark_EvaluationV1AndV2(b *testing.B) {
+	s := new(DBTestSuite)
+
+	t := &testing.T{}
+	s.SetT(t)
+	s.SetupSuite()
+
+	server := server.New(zap.NewNop(), s.store)
+	importer := ext.NewImporter(server)
+	reader, err := os.Open("./testdata/benchmark_test.yml")
```
Comment (markphelps) [inline internal/storage/sql/evaluation_test.go:705]: should we defer close this file?

**Review by markphelps — COMMENTED** [review]
looks good, one request and one or two questions for you and @GeorgeMac

File: internal/server/evaluation/server.go [hunk 2]
```diff
@@ -5,7 +5,7 @@ import (
 
 	"go.flipt.io/flipt/internal/storage"
 	"go.flipt.io/flipt/rpc/flipt"
-	rpcEvalution "go.flipt.io/flipt/rpc/flipt/evaluation"
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/server/evaluation/server.go:8]: Oh this was my suggestion in Yoofis last PR. I usually just lowercase the lot with Go.

File: internal/storage/sql/evaluation_test.go [hunk 3]
```diff
@@ -682,3 +692,95 @@ func (s *DBTestSuite) TestGetEvaluationRollouts_NonDefaultNamespace() {
...
+	}
+
+	eserver := evaluation.New(zap.NewNop(), s.store)
+
+	b.ResetTimer()
+
+	for _, flagKey := range flagKeys {
+		b.Run(fmt.Sprintf("evaluation-v1-%s", flagKey), func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				evaluation, err := server.Evaluate(context.TODO(), &flipt.EvaluationRequest{
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/storage/sql/evaluation_test.go:742]: Also too, beyond this, its probably wise to not create a random condition inside the benchmark. (as in the segment name).
I think you ideally want the function to be as idempotent as possible.
Even if that is just randomly picking a segment before getting inside the `for b.N` loop.

[the remark this answers is not in the document]
Comment (yquansah) [reply internal/storage/sql/evaluation_test.go]: Thanks y'all. I pushed up another commit that addresses this.

File: internal/storage/sql/evaluation_test.go (hunk 1 above)
Comment (yquansah) [reply internal/storage/sql/evaluation_test.go:705]: @markphelps done!

File: internal/server/evaluation/server.go [hunk 4]
```diff
@@ -5,7 +5,7 @@ import (
...
+	rpcevaluation "go.flipt.io/flipt/rpc/flipt/evaluation"
```
Comment (markphelps) [inline internal/server/evaluation/server.go:8]:
```suggestion
	"go.flipt.io/flipt/rpc/flipt/evaluation"
```

can we just use `evaluation` like we do `flipt` for the `go.flipt.io/flipt/rpc/flipt` import?

File: internal/storage/sql/evaluation_test.go [hunk 5]
```diff
@@ -682,3 +692,95 @@ func (s *DBTestSuite) TestGetEvaluationRollouts_NonDefaultNamespace() {
...
+
+	err = importer.Import(context.TODO(), reader)
+	require.NoError(t, err)
+
+	flagKeys := make([]string, 0, 10)
+
+	for i := 0; i < 10; i++ {
+		var flagKey string
+
+		num := rand.Intn(50)
```
Comment (markphelps) [inline internal/storage/sql/evaluation_test.go:720]:
```suggestion
		num := rand.Intn(50)+1
```

nit: would defeat the need of doing the `if num == 0` check below

**Review by markphelps — APPROVED** [review]
looks great, two minor nits

**Comment by markphelps** [comment]
sidenote, its interesting that boolean eval is so much faster and does 10x less allocs than variant eval:

```
Benchmark_EvaluationV1AndV2/evaluation-v1-flag_001#01-10      	     745	   1551987 ns/op	  148599 B/op	    6065 allocs/op
Benchmark_EvaluationV1AndV2/variant-evaluation-flag_001#01-10 	     740	   1568466 ns/op	  170978 B/op	    6157 allocs/op
Benchmark_EvaluationV1AndV2/boolean-evaluation-flag_boolean-10         	    2365	    489929 ns/op	   30777 B/op	     676 allocs/op
Benchmark_EvaluationV1AndV2/boolean-evaluation-another_boolean_flag-10 	    2409	    490192 ns/op	   30809 B/op	     676 allocs/op
```

**Comment by GeorgeMac** [comment]
> sidenote, its interesting that boolean eval is so much faster and does 10x less allocs than variant eval:
> 
> ```
> Benchmark_EvaluationV1AndV2/evaluation-v1-flag_001#01-10      	     745	   1551987 ns/op	  148599 B/op	    6065 allocs/op
> Benchmark_EvaluationV1AndV2/variant-evaluation-flag_001#01-10 	     740	   1568466 ns/op	  170978 B/op	    6157 allocs/op
> Benchmark_EvaluationV1AndV2/boolean-evaluation-flag_boolean-10         	    2365	    489929 ns/op	   30777 B/op	     676 allocs/op
> Benchmark_EvaluationV1AndV2/boolean-evaluation-another_boolean_flag-10 	    2409	    490192 ns/op	   30809 B/op	     676 allocs/op
> ```

In this case pprof would be more insightful as to where this is all accumulating.
One things to consider is that old eval has to do both rule and distribution lookup queries.
The big join Yoofi is doing means all the state is allocated in one shot.

That said, one thing you can't see with Go profiling tools is the memory profile in CGo land.
You can see the result that is scanned back into Go, but all the memory allocated for SQLite internals wont show up here.
Then again, you put Postgres and MySQL behind this and its a different story again.

**Pull request merged**